ovs-dpdk.md

```diff
@@ -0,0 +1,180 @@
+```
+# lspci | grep Ether
+```
+Review the status of current port bindings:
+```
+# driverctl -v list-devices
+```
+Bind the device to DPDK:
+```
+# driverctl set-override <PCI_ID> vfio-pci
+```
+
+### Hugepages Support
+Hugepages support is required for increasing performance, they allow the Linux kernel to manage larger pages, in addition to the standard 4K pages.
+Modern cpus use page-based mechanism to translate virtual address to physical address. The mapping is stored in page tables and can grow significantly large and to make the translation faster we use a transparant cache called TLB (Translation Lookaside Buffer). TLB miss will be handled by the hardware, which will increase the translation time significantly. TLB misses can be reduced by increasing the page size. Therefore, 1GiB pages are preferable.
+
+Number of pages can vary according to usage requirements, but is limited to memory size.
+Hugepages must be marked as shared in order to enable host and guest communication, as packets are transmitted and received on buffers allocated on hugepages memory. 
+
+Please note that 1Gib hugepages feature is not supported on every CPU 
+
+Review custom properties
+```
+# engine-config -g UserDefinedVMProperties
+```
+Add engine custom properties for hugepages support (followed by engine restart)
+```
+# engine-config -s "UserDefinedVMProperties=hugepages=^.*$;hugepages_shared=(true|false)"
+```
+ 
+
+### Enable DPDK in OVS:
+
+ovs-vscwhitchd should get a few parameters in order to enable DPDK functionality.
+
+* Set ```dpdk-init``` so OVS would initialize and support DPDK ports:
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-init=true
+```
+ 
+
+* ```dpdk-socket-mem```: a comma-separated list of amount of memory to allocate from each NUMA node. It is important to allocate hugepage memory to all NUMA nodes that have DPDK interfaces associated with them, otherwise, the NUMA node cannot be used and errors will occur. 
+For example, in a dual socket system, we allocate 1GiB from NUMA node 0 and 0GiB from the second:
+
+```
+# ovs-vsctl --no-wait set Open_vSwitch . other_config:dpdk-socket-mem="1024,0" 
+```
+
+* ```pmd-cpu-mask```: PMD (poll mode driver) threads are used to poll DPDK devices for new packets instead of using using the NIC driver that interacts with the kernel. 
+To better scale the workloads across cores, multiple pmd threads can be created and pinned to CPU cores by explicitly specifying pmd-cpu-mask. Ensure that cores from the same NUMA nodes as the network device are enabled. In case of Hyper Threading, it is best to use sibling cpus.
+
+eg: To spawn 2 pmd threads and pin them to cores 1, 2:
+
+```
+# ovs-vsctl set Open_vSwitch . other_config:pmd-cpu-mask=6
+
+```
+
+* ```dpdk-lcore-mask```: Specifies the CPU cores on which dpdk lcore threads should be spawned and expects hex string. These are threads for OVS usage other than PMD. They cannot be same as PMD or vCPUs threads.
```

Comment:
Suggest:
"DPDK"
"and expects a hex string."
"They cannot be the same as PMD or vCPUs threads."